src/main/cpp/src/datetime_parser.cu

```diff
@@ -0,0 +1,560 @@
+          if (!is_valid_digits(i, current_segment_digits)) {
+            return thrust::make_tuple(error_us, false);
+          }
+          segments[i]            = current_segment_value;
+          current_segment_value  = 0;
+          current_segment_digits = 0;
+          i += 1;
+        } else {
+          return thrust::make_tuple(error_us, false);
+        }
+      } else if (3 == i || 4 == i) {
+        if (':' == b) {
+          if (!is_valid_digits(i, current_segment_digits)) {
+            return thrust::make_tuple(error_us, false);
+          }
+          segments[i]            = current_segment_value;
+          current_segment_value  = 0;
+          current_segment_digits = 0;
+          i += 1;
+        } else {
+          return thrust::make_tuple(error_us, false);
+        }
+      } else if (5 == i || 6 == i) {
+        if ('.' == b && 5 == i) {
+          if (!is_valid_digits(i, current_segment_digits)) {
+            return thrust::make_tuple(error_us, false);
+          }
+          segments[i]            = current_segment_value;
+          current_segment_value  = 0;
+          current_segment_digits = 0;
+          i += 1;
+        } else {
+          if (!is_valid_digits(i, current_segment_digits)) {
+            return thrust::make_tuple(error_us, false);
+          }
+          segments[i]            = current_segment_value;
+          current_segment_value  = 0;
+          current_segment_digits = 0;
+          i += 1;
+          tz = cudf::string_view(bytes + j, (bytes_length - j));
+          j  = bytes_length - 1;
+        }
+        if (i == 6 && '.' != b) { i += 1; }
+      } else {
+        if (i < segments_len && (':' == b || ' ' == b)) {
+          if (!is_valid_digits(i, current_segment_digits)) {
+            return thrust::make_tuple(error_us, false);
+          }
+          segments[i]            = current_segment_value;
+          current_segment_value  = 0;
+          current_segment_digits = 0;
+          i += 1;
+        } else {
+          return thrust::make_tuple(error_us, false);
+        }
+      }
+    } else {
+      if (6 == i) { digits_milli += 1; }
+      // We will truncate the nanosecond part if there are more than 6 digits, which results
+      // in loss of precision
+      if (6 != i || current_segment_digits < 6) {
+        current_segment_value = current_segment_value * 10 + parsed_value;
+      }
+      current_segment_digits += 1;
+    }
+    j += 1;
+  }
+
+  if (!is_valid_digits(i, current_segment_digits)) { return thrust::make_tuple(error_us, false); }
+  segments[i] = current_segment_value;
+
+  while (digits_milli < 6) {
+    segments[6] *= 10;
+    digits_milli += 1;
+  }
+
+  cudf::string_view time_zone;
+  if (tz.has_value()) {
+    time_zone = tz.value();
+  } else {
+    time_zone = cudf::string_view(default_time_zone, default_time_zone_char_len);
+  }
+
+  segments[0] *= year_sign.value_or(1);
+  // above is ported from Spark.
+
+  // set components
```

Comment:
Could you add a comment about where this code comes from? It does not match 

https://github.com/apache/spark/blob/v3.5.0/sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala#L394

Reply:
Added comments.